src/rlang/vec-lgl.c

```diff
@@ -35,60 +36,71 @@ r_obj* r_lgl_which(r_obj* x, bool na_propagate) {
   const r_ssize n = r_length(x);
   const int* v_x = r_lgl_cbegin(x);
 
-  r_obj* names = r_names(x);
-  const bool has_names = names != r_null;
-  r_obj* const* v_names = NULL;
-  if (has_names) {
-    v_names = r_chr_cbegin(names);
-  }
-
-  const r_ssize which_n = r_lgl_sum(x, na_propagate);
+  const r_ssize out_n = r_lgl_sum(x, na_propagate);
 
-  if (which_n > INT_MAX) {
+  if (out_n > INT_MAX) {
     r_stop_internal("Can't fit result in an integer vector.");
   }
 
-  r_obj* which = KEEP(r_alloc_integer(which_n));
-  int* v_which = r_int_begin(which);
-
-  r_obj* which_names = r_null;
-  if (has_names) {
-    which_names = r_alloc_character(which_n);
-    r_attrib_poke_names(which, which_names);
-  }
+  r_obj* out = KEEP(r_alloc_integer(out_n));
+  int* v_out = r_int_begin(out);
 
-  r_ssize j = 0;
+  r_obj* names = r_names(x);
+  const bool has_names = (names != r_null);
 
   if (na_propagate) {
-    for (r_ssize i = 0; i < n; ++i) {
-      const int elt = v_x[i];
-
-      if (elt != 0) {
-        v_which[j] = (elt == r_globals.na_lgl) ? r_globals.na_int : i + 1;
-
-        if (has_names) {
-          r_chr_poke(which_names, j, v_names[i]);
-        }
-
-        ++j;
+    if (has_names) {
+      // Mark `NA` locations with negative location for extracting names later
+      for (r_ssize i = 0, j = 0; i < n && j < out_n; ++i) {
+        const int x_elt = v_x[i];
+        const bool missing = x_elt == r_globals.na_lgl;
+        const int elt = missing * (-i - 1) + !missing * x_elt * (i + 1);
+        v_out[j] = elt;
+        j += (bool) elt;
+      }
+    } else {
+      for (r_ssize i = 0, j = 0; i < n && j < out_n; ++i) {
+        const int x_elt = v_x[i];
+        const bool missing = x_elt == r_globals.na_lgl;
+        const int elt = missing * r_globals.na_int + !missing * x_elt * (i + 1);
+        v_out[j] = elt;
+        j += (bool) elt;
       }
     }
   } else {
-    for (r_ssize i = 0; i < n; ++i) {
-      const int elt = v_x[i];
-
-      if (elt == 1) {
-        v_which[j] = i + 1;
-
-        if (has_names) {
-          r_chr_poke(which_names, j, v_names[i]);
-        }
+    for (r_ssize i = 0, j = 0; i < n && j < out_n; ++i) {
+      const int x_elt = v_x[i];
+      v_out[j] = i + 1;
+      j += (x_elt == 1);
+    }
+  }
 
-        ++j;
+  if (has_names) {
```

Comment:
Names handling has been moved out of the core loop

Since our output here is a `which()` result, it points directly to locations in the original vector to indicate where we should pull the names from. This is great because it means we don't have to iterate through all of `x` again, we just do a fast loop through `v_out`

It gets a little complicated when there are names and `na_propagate = true`, but the idea is the same